upup/pkg/fi/cloudup/scaleway/cloud.go

```diff
@@ -16,7 +16,393 @@ limitations under the License.
+	panic("implement me")
+}
+
+func (s *scwCloudImplementation) GetClusterServers(clusterName string, serverName *string) ([]*instance.Server, error) {
+	request := &instance.ListServersRequest{
+		Zone: s.zone,
+		Name: serverName,
+		Tags: []string{TagClusterName + "=" + clusterName},
+	}
+	servers, err := s.instanceAPI.ListServers(request, scw.WithAllPages())
+	if err != nil {
+		if serverName != nil {
+			return nil, fmt.Errorf("failed to list cluster servers named %q: %w", *serverName, err)
+		}
+		return nil, fmt.Errorf("failed to list cluster servers: %w", err)
+	}
+	return servers.Servers, nil
+}
+
+func (s *scwCloudImplementation) GetClusterVolumes(clusterName string) ([]*instance.Volume, error) {
+	volumes, err := s.instanceAPI.ListVolumes(&instance.ListVolumesRequest{
+		Zone: s.zone,
+		Tags: []string{TagClusterName + "=" + clusterName},
+	}, scw.WithAllPages())
+	if err != nil {
+		return nil, fmt.Errorf("failed to list cluster volumes: %w", err)
+	}
+	return volumes.Volumes, nil
+}
+
+func (s *scwCloudImplementation) DeleteServer(server *instance.Server) error {
+	srv, err := s.instanceAPI.GetServer(&instance.GetServerRequest{
+		Zone:     s.zone,
+		ServerID: server.ID,
+	})
+	if err != nil {
+		if is404Error(err) {
+			klog.V(4).Infof("delete server %s: instance was already deleted", server.ID)
+			return nil
+		}
+		return err
+	}
+
+	// If the server is running, we turn it off and wait before deleting it
+	if srv.Server.State == instance.ServerStateRunning {
+		_, err := s.instanceAPI.ServerAction(&instance.ServerActionRequest{
+			Zone:     s.zone,
+			ServerID: server.ID,
+			Action:   instance.ServerActionPoweroff,
+		})
+		if err != nil {
+			return fmt.Errorf("delete server %s: error powering off instance: %w", server.ID, err)
+		}
+	}
+	_, err = s.instanceAPI.WaitForServer(&instance.WaitForServerRequest{
+		ServerID: server.ID,
+		Zone:     s.zone,
+	})
+	if err != nil {
+		return fmt.Errorf("delete server %s: error waiting for instance after power-off: %w", server.ID, err)
+	}
+
+	// We delete the server and wait before deleting its volumes
+	err = s.instanceAPI.DeleteServer(&instance.DeleteServerRequest{
+		ServerID: server.ID,
+		Zone:     s.zone,
+	})
+	if err != nil {
+		return fmt.Errorf("delete server %s: error deleting instance: %w", server.ID, err)
+	}
+	for {
+		_, err := s.instanceAPI.GetServer(&instance.GetServerRequest{
+			Zone:     s.zone,
+			ServerID: server.ID,
+		})
+		if is404Error(err) {
+			break
+		}
+	}
+
+	// We delete the volumes that were attached to the server (including etcd volumes)
+	for i := range server.Volumes {
+		err = s.instanceAPI.DeleteVolume(&instance.DeleteVolumeRequest{
+			Zone:     s.zone,
+			VolumeID: server.Volumes[i].ID,
+		})
+		if err != nil {
+			return fmt.Errorf("delete server %s: error deleting volume %s: %w", server.ID, server.Volumes[i].Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (s *scwCloudImplementation) DeleteVolume(volume *instance.Volume) error {
```

Comment:
Should you add a wait for Volume at the end of this function to be sure to return when you are in a safe state?